File: 프로그래머스/3/132266. 부대복귀/부대복귀.java
// 1. 인접 리스트 생성
// 2. 도착지에서 모든 노드로의 최단 거리 계산
// 3. sources 순회, 각 요소가 도착지에 도달할 수 있는 최단 거리를 answer에 push
import java.util.*;

class Solution {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        int[] answer = new int[sources.length];
        List<List<Integer>> adjList = new ArrayList<>();    // 2차원 리스트
        
        for (int i = 0; i <= n; i++) {
            adjList.add(new ArrayList<>()); // 각 행에 배열 넣기
        }
        
        for (int[] road: roads) {
            int a = road[0];
            int b = road[1];
            
            adjList.get(a).add(b);
            adjList.get(b).add(a);
        }
        
        Queue<Integer> queue = new ArrayDeque<>();
        boolean[] visited = new boolean[n + 1];
        int[] dist = new int[n + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        
        queue.add(destination);
        visited[destination] = true;
        dist[destination] = 0;
        
        while (!queue.isEmpty()) {
            int curr = queue.poll();
            
            for (int adj: adjList.get(curr)) {
                if (!visited[adj]) {
                    queue.add(adj);
                    visited[adj] = true;
                    dist[adj] = dist[curr] + 1;
                }
            }
        }
        
        for (int i = 0; i < sources.length; i++) {
            answer[i] = (dist[sources[i]]== Integer.MAX_VALUE) ? -1 : dist[sources[i]];
        }
         
        return answer;
    }
}

File: 프로그래머스/3/132266. 부대복귀/부대복귀.js
function solution(n, roads, sources, destination) {
    const answer = [];
    const adjList = Array.from({ length: n + 1 }, () => []); // 인접 리스트
    
    // 1. 인접 리스트 생성
    roads.forEach(([a, b]) => {
        adjList[a].push(b);
        adjList[b].push(a);
    })
    
    // destination에서 모든 노드로의 최단 거리 계산 (단일 BFS)
    const queue = [destination];
    const visited = new Array(n + 1).fill(false);
    const dist = new Array(n + 1).fill(Infinity);

    visited[destination] = true;
    dist[destination] = 0;

    while (queue.length > 0) {
        let curr = queue.shift();

        for (let adj of adjList[curr]) {
            if (!visited[adj]) {
                queue.push(adj);
                visited[adj] = true;
                dist[adj] = dist[curr] + 1;
            }
        }
    }
    
    // 2. source에서 destination까지의 최단 경로 구하고 answer에 push
    sources.forEach(source => {
        answer.push(dist[source] === Infinity ? -1 : dist[source]);
    })
    
    return answer;
}

// 시간 초과 발생 코드 -> 개선 방안: 매번 BFS 호출 X. 단일 BFS로 최단 거리를 모두 구하기
/*
function solution(n, roads, sources, destination) {
    const answer = [];
    const adjList = Array.from({ length: n + 1 }, () => []); // 인접 리스트
    
    // 1. 인접 리스트 생성
    roads.forEach(([a, b]) => {
        adjList[a].push(b);
        adjList[b].push(a);
    })
    
    const BFS = (start, destination) => {
        const queue = [];
        const visited = new Array(n + 1).fill(false);
        const dist = new Array(n + 1).fill(Infinity);
        
        queue.push(start);
        visited[start] = true;
        dist[start] = 0;
        
        while (queue.length > 0) {
            let curr = queue.shift();
            
            if (curr === destination) return dist[curr];
            
            for (let adj of adjList[curr]) {
                if (!visited[adj]) {
                    queue.push(adj);
                    visited[adj] = true;
                    dist[adj] = dist[curr] + 1;
                }
            }
        }
        
        return -1;
    }
    
    // 2. source에서 destination까지의 최단 경로 구하고 answer에 push
    sources.forEach(source => {
        answer.push(BFS(source, destination));
    })
    
    return answer;
}
*/